framework/src/base/Factory.C

```diff
@@ -62,6 +62,9 @@ Factory::create(const std::string & obj_name,
   // Print out deprecated message, if it exists
   deprecatedMessage(obj_name);
 
+  // Store the type
+  parameters.set<std::string>("type") = obj_name;
```

Comment:
yeah... Somehow we've managed to get away without explicitly storing this for 10 years. Not sure I want to add it now.